### PR TITLE
Fix bad unicode character issue

### PR DESCRIPTION
**Pull request by koechkevin**
## Description

- Fix an issue with our initial implementation of html2text where meedan returns bad unicode character error.
- Replaces `html2text` with `trafilatura`.
- A formatted input from [this post](https://pesacheck.org/contexte-manquant-cette-photo-ne-montre-pas-un-missile-sol-air-install%C3%A9-%C3%A0-goma-par-la-rdc-102461e13344) would now be processed as
[output.txt](https://github.com/CodeForAfrica/api/files/14523347/output.txt)

Fixes  [(issue)](https://code-for-africa.sentry.io/issues/5014867222/?environment=production&project=4506671517466624&query=&referrer=issue-stream&statsPeriod=14d&stream_index=0)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by kilemensi** [comment]
Before review the code, what is the actual issue here @koechkevin ? Meedan Check doesn't accept links?

**Comment by koechkevin** [comment]
@kilemensi How html2text handles some href links created some invalid unicode characters causing the error. Of all the options mentioned in the [documentation](https://github.com/Alir3z4/html2text/blob/master/docs/usage.md), ignore_links is the only that seemed to work for the specific case. Maybe another edge case will be introduced with other uploads

**Comment by kilemensi** [comment]
The option may work but I still don't understand the issue @koechkevin: Why would links raise unicode error? May be there is actually **unicode** error in the original article itself?

1. Does Meedan Check expect `unicode` or `ascii`?
2. Is there a maximum length limit?


If Meedan Check accepts unicode,
1. Have we tried the `unicode_snob` option?
2. If this didn't work, did we try manually encoding e.g. `html2text(html).encode("utf-8", errors="replace").decode("utf-8", errors="ignore")`?

**Comment by koechkevin** [comment]
@kilemensi this is the encoded url that causes the issues after parse by html2text https://www.reuters.com/world/middle-east/egypt-steps-up-security-border-israeli-offensive-gaza-nears-2024-02-09/#:~:text=CAIRO%2C%20Feb%209%20\(Reuters\),two%20Egyptian%20security%20sources%20said.

**Comment by kilemensi** [comment]
> @kilemensi this is the encoded url that causes the issues after parse by html2text [https://www.reuters.com/world/middle-east/egypt-steps-up-security-border-israeli-offensive-gaza-nears-2024-02-09/#:~:text=CAIRO%2C%20Feb%209%20\(Reuters\),two%20Egyptian%20security%20sources%20said](https://www.reuters.com/world/middle-east/egypt-steps-up-security-border-israeli-offensive-gaza-nears-2024-02-09/#:~:text=CAIRO%2C%20Feb%209%20%5C(Reuters%5C),two%20Egyptian%20security%20sources%20said).

But this doesn't seem to have any unicode issues @koechkevin ... as a matter of fact, it's actually just ascii. How does it look after being parsed by html2text?

**Comment by koechkevin** [comment]
> > @kilemensi this is the encoded url that causes the issues after parse by html2text [https://www.reuters.com/world/middle-east/egypt-steps-up-security-border-israeli-offensive-gaza-nears-2024-02-09/#:~:text=CAIRO%2C%20Feb%209%20(Reuters),two%20Egyptian%20security%20sources%20said](https://www.reuters.com/world/middle-east/egypt-steps-up-security-border-israeli-offensive-gaza-nears-2024-02-09/#:~:text=CAIRO%2C%20Feb%209%20%5C(Reuters%5C),two%20Egyptian%20security%20sources%20said).
> 
> But this doesn't seem to have any unicode issues @koechkevin ... as a matter of fact, it's actually just ascii. How does it look after being parsed by html2text?
Yes, no unicode issue @kilemensi but html2text adds escape characters around brackets on`(Reuters)` which is rejected by meedan.

**Comment by koechkevin** [comment]
[Egypt](https://www.reuters.com/world/middle-east/egypt-steps-up-security-\nborder-israeli-offensive-gaza-\nnears-2024-02-09/#:~:text=CAIRO%2C%20Feb%209%20\\(Reuters\\),two%20Egyptian%20security%20sources%20said.)

**Comment by koechkevin** [comment]
@kilemensi based on [caiosba's response](https://code4africa.slack.com/archives/C8WCUQ5C2/p1709256884845349?thread_ts=1697006394.382589&cid=C8WCUQ5C2), I suggest we go with this implementation because html2text implements links as markdown

**Comment by kilemensi** [comment]
> @kilemensi based on [caiosba's response](https://code4africa.slack.com/archives/C8WCUQ5C2/p1709256884845349?thread_ts=1697006394.382589&cid=C8WCUQ5C2), I suggest we go with this implementation because html2text implements links as markdown

I think caiosba's saying links are important @koechkevin :
> for the URLs, since those can be sent to end users on the tiplines, I suggest you simplify them and remove the fragments, in order to guarantee that WhatsApp linkify them correctly
>

If by current implementation you mean removing links then doesn't that mean URLs won't be sent to the end users?

**Comment by koechkevin** [comment]
> > @kilemensi based on [caiosba's response](https://code4africa.slack.com/archives/C8WCUQ5C2/p1709256884845349?thread_ts=1697006394.382589&cid=C8WCUQ5C2), I suggest we go with this implementation because html2text implements links as markdown
> 
> I think caiosba's saying links are important @koechkevin :
> 
> > for the URLs, since those can be sent to end users on the tiplines, I suggest you simplify them and remove the fragments, in order to guarantee that WhatsApp linkify them correctly
> 
> If by current implementation you mean removing links then doesn't that mean URLs won't be sent to the end users?

@kilemensi  yes

**Comment by koechkevin** [comment]
@kilemensi I have updated this PR.

As per [this](https://code4africa.slack.com/archives/C8WCUQ5C2/p1709647081764739?thread_ts=1709619656.171799&cid=C8WCUQ5C2),

Links are useful and content should be plaintext (No markdown). However, since html2text only converts to markdown, I have used [Beautiful Soup](https://beautiful-soup-4.readthedocs.io/en/latest/) to convert this to text appended links(href & img src) so we won't be having bad unicode errors anymore.

**Comment by kilemensi** [comment]
> Links are useful and content should be plaintext (No markdown). However, since html2text only converts to markdown, I have used [Beautiful Soup](https://beautiful-soup-4.readthedocs.io/en/latest/) to convert this to text appended links(href & img src) so we won't be having bad unicode errors anymore.
>

`beatufulsoup` output isn't very good @koechkevin... lets just pull [the big gun](https://github.com/adbar/trafilatura?tab=readme-ov-file#evaluation-and-alternatives) out and use `trafilatura` then.

**Comment by koechkevin** [comment]
> > Links are useful and content should be plaintext (No markdown). However, since html2text only converts to markdown, I have used [Beautiful Soup](https://beautiful-soup-4.readthedocs.io/en/latest/) to convert this to text appended links(href & img src) so we won't be having bad unicode errors anymore.
> 
> `beatufulsoup` output isn't very good @koechkevin... lets just pull [the big gun](https://github.com/adbar/trafilatura?tab=readme-ov-file#evaluation-and-alternatives) out and use `trafilatura` then.

@kilemensi this is updated

**Comment by kilemensi** [comment]
> @kilemensi this is updated
>

Can you include an output sample in the PR description @koechkevin ?

**Comment by kilemensi** [comment]
Is output.txt the current output @koechkevin? Asking because I see links so not sure if this is better than html2text or not.

**Comment by koechkevin** [comment]
> Is output.txt the current output @koechkevin? Asking because I see links so not sure if this is better than html2text or not.

@kilemensi Yes, output.txt is the current implementation. It is better than html2text because, it does not process the whole html as a markdown but only parts with the links are formatted like markdown and no bad unicode characters included.

File: pesacheck_meedan_bridge/py/main.py [hunk 1]
```diff
@@ -6,7 +6,15 @@
...
 from check_api import post_to_check
 from database import PesacheckDatabase, PesacheckFeed
-from html2text import html2text
+from lxml import html
+from trafilatura import extract
+
+
+def html_to_formatted_text(content):
+    tree = html.fromstring(content)
+    text = extract(tree, include_links=True, include_images=True)
```
[the remark this answers is not in the document]
Comment (koechkevin) [reply pesacheck_meedan_bridge/py/main.py:15]: It doesn't, extract as string returns empty string

[the remark this answers is not in the document]
Comment (kilemensi) [reply pesacheck_meedan_bridge/py/main.py:15]: Doc says it does... unless it's a bug

> Parameters:
> - filecontent – HTML code as string.

File: 3rdparty/py/requirements-all.txt [hunk 2]
```diff
@@ -13,8 +13,10 @@ google-auth-oauthlib==1.2.0
 greenlet==3.0.3
 gunicorn[gevent, setproctitle]==21.2.0
 html2text==2024.2.26
+lxml==5.1.0
```
Comment (kilemensi) [inline 3rdparty/py/requirements-all.txt:16]: I'm assuming we also don't need this, right?

**Review by kilemensi — APPROVED** [review]

**Pull request merged**